src/internal/events/events.go

```diff
@@ -159,3 +162,84 @@ func tenantHash(tenID string) string {
 	sum := md5.Sum([]byte(tenID))
 	return fmt.Sprintf("%x", sum)
 }
+
+// ---------------------------------------------------------------------------
+// metrics aggregation
+// ---------------------------------------------------------------------------
+
+type m string
+
+// metrics collection bucket
+const APICall m = "api_call"
+
+// configurations
+const (
+	reportInterval    = 1 * time.Minute
+	retentionDuration = 2 * time.Minute
+)
+
+// NewMetrics embeds a metrics bus into the provided context.  The bus can be
+// utilized with calls like Inc and Since.
+func NewMetrics(ctx context.Context, w io.Writer) (context.Context, func()) {
+	var (
+		// report interval time-bounds metrics into buckets.  Retention
+		// controls how long each interval sticks around.  Neither one controls
+		// logging rates; that's handled by dumpMetrics().
+		sink = metrics.NewInmemSink(reportInterval, retentionDuration)
+		cfg  = metrics.DefaultConfig("corso")
+		sig  = metrics.NewInmemSignal(sink, metrics.DefaultSignal, w)
+	)
+
+	cfg.EnableHostname = false
+	cfg.EnableRuntimeMetrics = false
+
+	gm, err := metrics.NewGlobal(cfg, sink)
+	if err != nil {
+		logger.CtxErr(ctx, err).Error("metrics bus constructor")
+		return ctx, func() { sig.Stop() }
+	}
+
+	stop := make(chan struct{})
+	go dumpMetrics(ctx, stop, sig)
+
+	flush := func() {
+		close(stop)
+		sig.Stop()
+		gm.Shutdown()
+	}
+
+	// return context.WithValue(ctx, sinkCtxKey, sink), flush
+	return ctx, flush
+}
+
+// dumpMetrics runs a loop that sends a os signal (SIGUSR1 on linux/mac, SIGBREAK on windows)
+// every logging interval.  This syscall getts picked up by the metrics inmem signal and causes
+// it to dump metrics to the provided writer (which should be the logger).
+// Expectation is for users to call this in a goroutine.  Any signal or close() on the stop chan
+// will exit the loop.
+func dumpMetrics(ctx context.Context, stop <-chan struct{}, sig *metrics.InmemSignal) {
+	tock := time.Tick(reportInterval)
+
+	for {
+		select {
+		case <-tock:
+			if err := syscall.Kill(syscall.Getpid(), metrics.DefaultSignal); err != nil {
+				logger.CtxErr(ctx, err).Error("metrics interval signal")
+			}
+		case <-stop:
+			return
+		}
+	}
+}
+
+// Inc increments the given category by 1.
+func Inc(cat m, keys ...string) {
+	cats := append([]string{string(cat)}, keys...)
+	metrics.IncrCounter(cats, 1)
+}
+
+// Since records the duration between the provided time and now, in millis.
```

Comment:
I'm confused by this comment and how this function is called in `service.go`. If we're getting the duration between the passed in time and the current time and all of this is done in the `Intercept()` function, is it really netting us useful metrics? Won't it just be the few ms it takes to generate the original timestamp and then reach this function call?

Reply:
> is it really netting us useful metrics? Won't it just be the few ms it takes to generate the original timestamp and then reach this function call?

The implementation is effectively:
```go
	start     := time.Now()
	resp, err := pipeline.Next(req, middlewareIndex)
	events.Since(start, events.APICall)
```
The sequence of events will work as follows:
1. start the timer.
2. call pipeline.Next.  This will call the downstream http RoundTripper (ie, make the http request), blocking until it gets resolved.
3. call Since() to mark the end of the timer.  The recorded duration is `now() - start`: ie, the difference in time between 1/ and 3/.

The result is a measurement of the http RoundTripper (and any other middleware stacked on this one, so it's important for this to sit at the top).

Reply:
thanks for pointing that out, I skimmed over the `pipeline.Next()` call because I'm not used to stuff doing work in `var` blocks 😅

Reply:
hah, yeah, it does feel a little weird.